src/main/java/io/wgj/ctci/Stack.java

```
@@ -1,21 +1,80 @@
 package io.wgj.ctci;
 
-public class Stack {
-    int[] stack;
+/**
+ * Stack is a stack data structure backed by a dynamic array. Stack can be
+ * initialized with a starting array size. It's written to develop familiarity
+ * with the stack data structure, and should otherwise not be used.
+ * Maximum allowed array is 1024. If you set a starting capacity larger than
+ * this, the maximum allowed array size is used as the starting capacity. An
+ * UnsupportedOperationException will be thrown when more the maximum allowed
+ * amount of elements are added.
+ */
+public class Stack<E> {
+    static final int MAX_ARRAY_SIZE = 1024;
+    static final int DEFAULT_CAPACITY = 10;
+    Object[] stack;
+    int size;
     int position;
 
-    public Stack(){
-        stack = new int[10];
+    public Stack() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public Stack(int size) {
+        stack = new Object[size];
+        if (size > MAX_ARRAY_SIZE) {
+            size = MAX_ARRAY_SIZE;
+        }
+        this.size = size;
         position = 0;
     }
 
-    public void push(int i) {
-        stack[position] = i;
+    public void push(E e) {
+        if (size == position) {
+            this.grow();
+        }
+        stack[position] = e;
         position++;
     }
 
-    public int pop(){
+    public E pop() {
         position--;
-        return stack[position];
+        @SuppressWarnings("unchecked")
+        E e = (E) stack[position];
+        return e;
+    }
+
+    private void grow() {
+        int newSize = size << 2;
```

I think maybe you meant to multiply by 2?  "<< 2" effectively means "multiply by 4", which is more than necessary to get amortized linear complexity.  It will still work fine, but in practice will lean more heavily on memory usage.

Also you might disagree, but imho the only reason to use bit shifting in java in 2020 is to let people know you liked computers before nerds were cool.  It is unarguably more performant but it just doesn't matter.  It also handles overflows by silently failing (it is NOT multiplication), which might be bad.

---

Thought it was a cool way to grow. I stole `<<2` from OpenJDK's implementation of `ArrayList`, which being a major implementation of the API, can obviously benefit from the performance savings.